### PR TITLE
fix(catalog): resilient Logo render for malformed S3 URLs

### PR DESCRIPTION
**Pull request by fiskus**
## Summary
- `<Logo>` falls back to the default Quilt logo (and reports to Sentry) when the configured `src` is a malformed `s3://` URL, instead of crashing the render tree via `parseS3Url` throwing.
- New `s3paths.tryParseS3Url` — a null-returning, render-safe counterpart to `parseS3Url`.
- New `validators.urlWithHost` — stricter sibling of `validators.url` that rejects URLs with no hostname (e.g. `s3://`, which `new URL()` accepts because non-special schemes are allowed empty hosts). Used in the logo URL field so partial inputs like `s3://` are caught at validation time, not silently masked by `<Logo>`'s runtime fallback.
- Theme editor:
  - Adds `validators.urlWithHost` and `validators.s3Url` to the logo URL field's compose chain.
  - Surfaces validation errors as the user types (`meta.modified || meta.submitFailed`) instead of only after submit.
  - Extracts a small `Preview` sub-component; renders a `broken_image` placeholder for invalid URLs and a `hide_image` placeholder for the empty state.
  - Keeps the broken-image icon in sync with the helperText (both gated through the same `error` flag).

## Test plan
- [x] `npm run test:only -- validators Logo Admin/Settings/ThemeEditor` passes (14 + 5 + 4 tests).
- [ ] Manual: configure a custom logo with `s3://` (no bucket) — broken-image preview + "Enter a valid URL (https:// or s3://)" helper text.
- [ ] Manual: configure a valid `s3://bucket/key.png` — signed URL renders.
- [ ] Manual: configure a regular `https://` URL — renders unchanged.
- [ ] Manual: open the editor with a previously-saved malformed `s3://` value — catalog header gracefully falls back to the Quilt logo (no white screen, Sentry report fires); editor flags the value once the field is modified.

<h3>Greptile Summary</h3>

This PR adds two complementary layers of resilience for malformed S3 logo URLs: a runtime fallback in `<Logo>` that catches parse failures (including bucket-only URLs) and gracefully renders the default Quilt logo with a Sentry report, and a new `validators.logoUrl` that rejects no-hostname and no-key S3 URLs at form-input time.

- **`CustomLogo`** now wraps `parseS3Url` in a try/catch and checks for an empty key, returning a discriminated-union result; a `useEffect` reports errors to Sentry exactly once per distinct bad src.
- **`validators.logoUrl`** replaces the old `url`/`s3Url` pair with a single validator that uses `new window.URL(v).hostname` to reject empty-host URLs (e.g. `s3://`) and additionally rejects `s3://bucket` and `s3://bucket/` where no key is present.
- **ThemeEditor** switches error display to `meta.modified || meta.submitFailed` and introduces a `Preview` sub-component that gates the broken-image icon through the same `error` flag as `helperText`, keeping both UI signals in sync.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all changed paths have defence-in-depth (validator + runtime fallback) and are covered by new tests.

The change is well-scoped: it adds a try/catch + key-guard to one render function, a single new validator, and tightens one form field's error-display logic. Both the runtime fallback and the form-side validation are exercised by new tests covering the specific edge cases that triggered the original crash. No existing behaviour is removed in a way that could introduce regressions.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| catalog/app/components/Logo/index.tsx | Adds resilient error handling in CustomLogo: catches parseS3Url exceptions and rejects S3 URLs with empty keys, falling back to QuiltLogo and reporting to Sentry via a useEffect; uses a discriminated union type for type-safe branching. |
| catalog/app/utils/validators.js | Replaces the separate url/s3Url validators with a single logoUrl that checks hostname presence and requires a non-empty S3 key; correctly handles edge cases including s3://, s3://bucket, and s3://bucket/. |
| catalog/app/containers/Admin/Settings/ThemeEditor.tsx | Extracts a Preview sub-component that gates the broken_image icon through the same error flag as the helperText; changes error display from submitFailed-only to modified||submitFailed; switches to validators.logoUrl. |
| catalog/app/components/Logo/index.spec.tsx | Adds four new tests covering S3 logo signing, malformed s3:// fallback, bucket-only fallback, and Sentry reporting; mocks @sentry/react cleanly with a beforeEach reset. |
| catalog/app/utils/validators.spec.js | Adds a thorough logoUrl test suite covering empty values, valid URLs, malformed inputs, hostname-less URLs, and S3 URLs without keys. |

</details>

<sub>Reviews (3): Last reviewed commit: ["refactor(catalog): model &lt;Logo&gt; parse re..."](https://github.com/quiltdata/quilt/commit/0ee1dc17885c6b62fc732e314632792fd32b8ae1) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=31176108)</sub>

### CODE REVIEW

File: catalog/app/containers/Admin/Settings/ThemeEditor.tsx [hunk 1]
```diff
@@ -398,11 +405,13 @@ export default function ThemeEditor() {
...
+                        validators.s3Url,
                       ) as FF.FieldValidator<string>
                     }
                     errors={{
-                      url: 'Image should be a valid URL',
-                      file: 'Image should be file',
+                      url: 'Image should be a valid URL or S3 URL',
+                      s3Url: 'Image should be a valid URL or S3 URL',
+                      file: 'Image should be a file',
                     }}
```
[the remark this answers is not in the document]
Comment (fiskus) [reply catalog/app/containers/Admin/Settings/ThemeEditor.tsx:424]: Thanks — you're right that today, given `composeOr(file, url, s3Url)` returns the *last* validator's error on full failure, `errors.url` is unreachable.

Keeping the `url` key intentionally as a defensive fallback: `composeOr` itself has a UX smell ("last validator wins" is arbitrary — it surfaces `'s3Url'` for a string like `not a url` that any user would call a bad URL, not a bad S3 URL). When that gets fixed to surface a more sensible error (e.g., the *first* error, or the most specific one based on input shape), `errors.url` would start being reached for inputs that look more like a URL than an S3 URL. Cheaper to leave the message in place than to remember to add it back later.

File: catalog/app/containers/Admin/Settings/ThemeEditor.tsx [hunk 2]
```diff
@@ -140,28 +145,30 @@ export function InputFile({ input: { value, onChange }, meta, errors }: InputFil
...
+    setPreviewUrl(url)
     return () => URL.revokeObjectURL(url)
   }, [value])
-  const previewSrc = React.useMemo<string | null>(() => {
-    if (typeof value === 'string') return value || null
-    return blobUrl
-  }, [value, blobUrl])
-  const urlValue = typeof value === 'string' ? value : ''
+  const isUrl = typeof value === 'string' && value.length > 0
+  const isInvalidUrl = isUrl && !!meta?.error
```
[the remark this answers is not in the document]
Comment (fiskus) [reply catalog/app/containers/Admin/Settings/ThemeEditor.tsx]: Good catch — applied in 4bee37d3. The Preview component now reads from the gated `error` (which already requires `modified || submitFailed`) instead of raw `meta?.error`, so the broken-image icon and the helperText are now in sync. Saved-invalid URLs still render gracefully via `<Logo>`'s runtime fallback (QuiltLogo + Sentry), so nothing actually breaks visually for the user — they just don't see editor-side feedback until they interact.

**Comment by fiskus** [comment]
@greptileai please update confidence score

**Comment by fiskus** [comment]
@greptileai please update confidence score

**Comment by fiskus** [comment]
@greptileai please re-review

File: catalog/app/utils/validators.js [hunk 3]
```diff
@@ -130,24 +128,21 @@ export const hexColor = (v) => {
...
-export const s3Url = (v) => {
-  if (!v) return undefined
-  try {
-    const { bucket, key } = s3paths.parseS3Url(v)
-    if (!bucket || !key) return 's3Url'
-  } catch (e) {
-    return 's3Url'
-  }
+  return undefined
 }
```
[the remark this answers is not in the document]
Comment (fiskus) [reply catalog/app/utils/validators.js]: Good catch — addressed in 75ebbb2a4. Two layers:

- `validators.logoUrl` (replaces use of `validators.url` here): rejects URLs without a hostname AND `s3://` URLs without a key. So `s3://bucket` and `s3://bucket/` are now form-rejected with the same helperText.
- `<Logo>` runtime guard: treats `parseS3Url` results with empty `key` as malformed → falls back to `QuiltLogo` + Sentry. This covers NavBar / pre-saved settings (the validator only protects form input, not values already persisted before this fix).

Spec updates in the same commit cover the new cases on both sides.

**Comment by fiskus** [comment]
@greptileai please update confidence score

**Comment by fiskus** [comment]
@greptileai please re-review

**Review by drernie — APPROVED** [review]
